899 updated MembershipInferenceBlackBox#infer to have CUDA support

evansakmar opened the pull request:
Signed-off-by: Evan Sakmar <sakmarev@ctc.com>

# Description

Added two helper methods to_cuda and from_cuda to assist in moving the tensors to and from CUDA.  Updated the logic to move those tensors when appropriate.

Fixes #899

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

I ran through each of these steps
- Ran through the notebooks/attack_membership_inference.ipynb notebook in a CUDA enable instance and compared it to the results of the notebook run on a CPU only instance.
- Ran  the membership inference tests on a CUDA enabled instance and CPU only instance and ensured that all tests passed.

**Test Configuration**:
- OS: Ubuntu 18.04
- CUDA: 10.1
- Python version: 3.6.9
- TensorFlow / Keras / PyTorch / MXNet version: TensorFlow 2.4, Keras 2.4.3 torch 1.6.0+cu92, MXNet 1.6.0

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
#- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
#- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

evansakmar commented:
I wasn't sure if the to_cuda and from_cuda methods should live in a helper class somewhere else but kept them in the file for now since I didn't see to_cuda used anywhere else in the code base.  I am open to feedback on that.

Also if there is anything else I could improve in it let me know

Thanks
Evan

beat-buesser commented:
Hi @evansakmar Thank you very much for finding bugs and especially for following up with pull requests to fix them, this is very much appreciated. I have changed the target branch from `main` to `dev_1.5.2` to include your updates in this week's 1.5.2 release. Could you please rebase your branch to `dev_1.5.2`?

Could you please try moving the to_cuda and from_cuda methods to `art.utils` which is currently the main place for utility methods in ART. This utils file is getting long, so we'll split it soon into specific modules.

I have asked @abigailgold for a review of the changes.

abigailgold commented:
@beat-buesser Yes, I agree with this change, and I agree with your comment that the to_cuda and from_cuda methods should move to art.utils. @evansakmar Please make sure to import them only within: if self.default_model (where the rest of the torch imports are located). Thanks!

evansakmar commented:
Will do 

Thank You!

evansakmar commented:
@abigailgold @beat-buesser 

I moved those methods to art.utils.  I added the Tensor type to the methods to follow what the other methods do.  Let me know if you want to keep it generic and avoid the `from torch import Tensor`.

Thanks
Evan

beat-buesser commented:
@evansakmar @abigailgold Thank you very much!
